match wildcards values

architkshk opened the pull request:
fixes #125 . @abhishek97 Please check.

championswimmer commented:
@abhishek97 this would do it ?

abhishek97 commented on src/oauth/oauthserver.js:
```diff
@@ -79,7 +79,9 @@ server.exchange(oauth.exchange.code(
             let callbackMatch = false
             for (url of client.callbackURL) {
-                if (redirectURI.startsWith(url)) callbackMatch = true
+                const exp = "[^\s]+";
```
Why not `[^\s]*` ? wildcards and be replaced by empty strings also.

abhishek97 commented:
@championswimmer This would do for matching, however, I am not sure 

- Is this the only place where we match `redirectUrl` with the client's saved `callbackUrl` CSV? We also have a `domain` field. I think that would also have a "check" of its own.

- Probably we need a similar treatment here too: https://github.com/architkshk/oneauth/blob/2b1dea7b238ef60423424afe1bf4f0b89fa9142f/src/oauth/oauthserver.js#L130 

- This custom matching logic can be refactored out in utils. Might help in reusablility and unit testing.

architkshk commented:
@abhishek97 
- I cannot find any check for domain field.
- Replaced the other place where we have to check callback URL.
- Moved the code to utils/wildcardmatch.js

abhishek97 commented:
@architkshk Have you tried it with a sample app?

@championswimmer We can merge this on staging and test it out.

architkshk commented:
@abhishek97 I've tried the code to check for various cases though I did not check for any sample app.

architkshk commented:
@abhishek97 I've made changes to the code and also tried my creating sample clients on localhost. It works fine. Please review.

architkshk commented:
@abhishek97 @championswimmer any updates?

architkshk commented:
@abhishek97 @championswimmer any updates?

abhishek97 commented:
@championswimmer Let's try this out? I can find no checks for domains either.

championswimmer commented:
@abhishek97 cool, will try and merge today/tomorrow

championswimmer commented on src/oauth/oauthserver.js:
```diff
@@ -78,10 +79,7 @@ server.exchange(oauth.exchange.code(
-            for (url of client.callbackURL) {
-                if (redirectURI.startsWith(url)) callbackMatch = true
-            }
+            let callbackMatch = matchURL(client.callbackURL,redirectURIredirectURI);
```
dafuq? why did you remove the for loop :|

architkshk replied:
figured out the error. it's a type, in line 82 I have written redirectURI 2 times. I am making changes and recreating a PR.

championswimmer commented:
reverted. the for loop removal fucked up the entire logic

architkshk commented:
@championswimmer what test did it failed?

championswimmer commented:
none of the explicit auths were working

architkshk commented:
Done #235  works fine now.

architkshk commented:
Done #284  works fine now. @championswimmer
